username_from_threepid/__init__.py

```diff
@@ -11,36 +11,121 @@
+        """Checks that the required fields are present and at a correct value, and
+        instantiates a UsernameEmailConfig.
+
+        Args:
+            config: The raw configuration dict.
+
+        Returns:
+            A UsernameEmailConfig generated from this configuration
+
+        Raises:
+            ConfigError if a required field is missing or at an illegal value.
+        """
+        if "threepid_to_use" not in config:
+            raise ConfigError('missing required configuration key: "threepid_to_use"')
+
+        if config["threepid_to_use"] not in ["email", "msisdn"]:
+            raise ConfigError(
+                '"threepid_to_use" can only be either "email" or "msisdn"'
+            )
+
+        return UsernameEmailConfig(**config)
+
+    async def set_username_from_threepid(
+        self,
+        auth_result: Dict[str, Any],
+        params: Dict[str, Any],
+    ) -> Optional[str]:
+        """Look at the provided UIA result and figure out a username from the 3PIDs
+        provided during registration.
+
+        Args:
+            auth_result: The UIA results.
+            params: The body of the /register request. Not used by this module.
+
+        Returns:
+            A username, or None if no 3PID was provided and "fail_if_not_found" is set to
+            False.
+
+        Raises:
+            RuntimeError if no 3PID was provided and "fail_if_not_found" is set to True.
+        """
+        desired_username: Optional[str] = None
+
+        if (
+            self._config.threepid_to_use == "email"
+            and LoginType.EMAIL_IDENTITY in auth_result
+        ):
+            address = auth_result[LoginType.EMAIL_IDENTITY]["address"]
+
+            # filter out invalid characters
+            filtered = filter(
+                lambda c: c in mxid_localpart_allowed_characters,
+                address.replace("@", "-").lower(),
+            )
+            desired_username = "".join(filtered)
+
+            # find a unique mxid for the account, suffixing numbers
+            # if needed
+            while True:
+                try:
+                    await self._api.check_username(desired_username)
+                    # if we got this far we passed the check.
+                    break
+                except SynapseError as e:
+                    if e.errcode == "M_USER_IN_USE":
+                        m = re.match(r"^(.*?)(\d+)$", desired_username)
+                        if m:
+                            desired_username = m.group(1) + str(int(m.group(2)) + 1)
+                        else:
+                            desired_username += "1"
+                    else:
+                        # something else went wrong.
+                        raise
+        elif (
+            self._config.threepid_to_use == "msisdn" and LoginType.MSISDN in auth_result
+        ):
+            desired_username = auth_result[LoginType.MSISDN]["address"]
```

Comment:
Why doesn't this call `check_username` and append digits at the end if necessary? Oh, I guess because it's a phone number and so isn't going to have clashes like emails?

Reply:
It probably should. This is lifted almost verbatim from the synapse-dinsic code linked to above, which is likely because Tchap only uses email for signups, so I guess this side didn't get much love.